File: cap_linsis/codes/python/metodo_de_jacobi/jacobi.py
from __future__ import division
import numpy as np
from numpy import linalg

def jacobi(A,b,x0,tol,N):
    #preliminares
    A = A.astype('double')
    b = b.astype('double')
    x0 = x0.astype('double')
    
    n=np.shape(A)[0]
    x = np.zeros(n)
    it = 0
    #iteracoes
    while (it < N):
        it = it+1
        #iteracao de Jacobi
        for i in np.arange(n):
            x[i] = b[i]
            for j in np.concatenate((np.arange(0,i),np.arange(i+1,n))):
                x[i] -= A[i,j]*x0[j]
            x[i] /= A[i,i]
        #tolerancia
        if (np.linalg.norm(x-x0,np.inf) < tol):
            return x
        #prepara nova iteracao
        x0 = np.copy(x)
    raise NameError('num. max. de iteracoes excedido.')

File: cap_linsis/codes/octave/metodo_de_jacobi/jacobi.m
function [x]=jacobi(A,b,x,tol,N)
  n = size(A,1);
  x0 = x;
  iter = 1;
  while (iter <= N)
    #iteracao Jacobi
    for i = 1:n
      x(i) = (b(i) - A(i,[1:i-1,i+1:n])*x0([1:i-1,i+1:n]))/A(i,i);
    endfor
    #condicao de parada
    if (norm(x-x0,'inf')<tol)
      return
    endif
    #prepara nova iteracao
    x0 = x;
    iter += 1;
  endwhile
  error('num. max. iter. excedido.')  
endfunction

